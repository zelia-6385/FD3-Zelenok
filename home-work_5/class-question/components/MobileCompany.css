.MobileCompany {
    width: 180px;
}

.MobileCompanyName {
    margin-top: 10px;
    font-weight: bold;
}

.MobileCompanyClients {
    margin: 10px 0;
    display: table;
    width: 100%;
    border: 1px solid black;
    background-color: yellowgreen;
}

.MobileClient {
    display: table-row;
}

.MobileClientFIO,
.MobileClientBalance {
    display: table-cell;
}