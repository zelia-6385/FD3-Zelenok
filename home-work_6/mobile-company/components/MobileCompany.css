.MobileCompany__table {
    width: 80%;
}

.MobileCompany__table th {
    text-align: left;
}