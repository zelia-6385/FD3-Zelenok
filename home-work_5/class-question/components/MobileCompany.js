import React from 'react';
import PropTypes from 'prop-types';
import deepEqual from 'deep-equal';

import MobileClient from './MobileClient';

import './MobileCompany.css';

class MobileCompany extends React.Component {

    static propTypes = {
        name: PropTypes.string.isRequired,
        clients: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.number.isRequired,
                fio: PropTypes.string.isRequired,
                balance: PropTypes.number.isRequired,
            })
        ),
    };

    state = {
        name: this.props.name,
        clients: this.props.clients,
    };

    // shouldComponentUpdate = (newProps, newState) => {
    //     return !(deepEqual(newProps, this.props) && deepEqual(newState, this.state));
    // };

    shouldComponentUpdate = (newProps, newState) => {
        if(newState.name != this.state.name) {
            console.log(this.state.name);
            console.log(newState.name);
            console.log(3);
            return true;
        }
        if(newState.clients.length != this.state.clients.length) {
            return true;
        }
        let clientsChanged = false;
        newState.clients.forEach( (newClient, i) => {
            let oldClient = this.state.clients[i];
            console.log(oldClient.balance);
            console.log(newClient.balance);
            if (oldClient.fio != newClient.fio) {
                clientsChanged = true;
            }
            // if(oldClient.balance != newClient.balance) {
            //     console.log(1);
            //     clientsChanged = true;
            // }
            let oldActive = oldClient.balance >= 0;
            let newActive = newClient.balance >= 0;
            if(oldActive != newActive) {
                clientChanged = true;
            }
        });
        if(clientsChanged) {
            console.log(2);
            return true;
        }
        return false; 
    }

    setName1 = () => {
        this.setState({name: 'MTC'});
    };

    setName2 = () => {
        this.setState({name: 'Velcome'});
    };

    setBalance = (clientId, newBalance) => {
        let newClients = [...this.state.clients];// копия массива клиентов
        newClients.forEach( (c, i) => {
            if (c.id == clientId) {
                // c.balance = newBalance;
                let newClient = {...c};//копия хэша изменившегося клиента
                newClient.balance = newBalance;
                newClients[i] = newClient
            }
        });
        this.setState({clients: newClients});
    };

    setBalance1 = () => {
        this.setBalance(105, 230);
    };

    setBalance2 = () => {
        this.setBalance(105, 250);
    };

    updateBalance = (clientId, delta) => {
        let newClients = this.state.clients;
        newClients.map( c => {
            if(c.id == clientId) {
                c.balance += delta;
            }
        });
        console.log(newClients);
        this.setState({clients: newClients});
    };

    updateBalance1 = () => {
        this.updateBalance(110, -1);
    };

    updateBalance2 = () => {
        this.updateBalance(110, 1);
    };

    render() {

        console.log("MobileCompany render");

        // let clientsCode = this.state.clients.map( c => 
        //     <MobileClient key={c.id} id={c.id} fio={c.fio} balance={c.balance}/>
        // );

        let clientsCode = this.state.clients.map( client => 
            <MobileClient key={client.id} info={client}/>
        );

        return (
            <div className='MobileCompany'>
                <input type="button" value="=MTC" onClick={this.setName1} />
                <input type="button" value="=Velcome" onClick={this.setName2} />
                <div className='MobileCompanyName'>
                    Компания &laquo;{this.state.name}&raquo;
                </div>
                <div className='MobileCompanyClients'>
                    {clientsCode}
                </div>
                <input type="button" value="Сидоров=230" onClick={this.setBalance1} />
                <input type="button" value="Сидоров=250" onClick={this.setBalance2} />
                <input type="button" value="Петров--" onClick={this.updateBalance1} />
                <input type="button" value="Петров++" onClick={this.updateBalance2} />
            </div>
        );
    }
}

export default MobileCompany;