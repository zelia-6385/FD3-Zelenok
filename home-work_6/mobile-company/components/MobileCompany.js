import React from 'react';
import PropTypes from 'prop-types';

import './MobileCompany.css';

import MobileClient from './MobileClient';
import MobileClientCard from './MobileClientCard';
import {clientEvents} from './events';

class MobileCompany extends React.PureComponent {

    static propTypes = {
        name: PropTypes.string.isRequired,
        clients: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.number.isRequired,
                info: PropTypes.shape({
                    fam: PropTypes.string.isRequired,
                    im: PropTypes.string.isRequired,
                    otch: PropTypes.string.isRequired,
                    balance: PropTypes.number.isRequired, 
                }),
            })
        ),
    };

    state = {
        name: this.props.name,
        clients: this.props.clients,
        clientsCopy: [],

        client: {},
        mode: null,
        lastId: null,

    };

    componentDidMount = () => {
        clientEvents.addListener('EDeleteClicked', this.deleteClient);
        clientEvents.addListener('EEditeClicked', this.showEditCard);
        clientEvents.addListener('ESaveClicked', this.editClient);
        clientEvents.addListener('ECancelClicked', this.hideCard);
        clientEvents.addListener('EAddClicked', this.createNewClient);
    };

    componentWillUnmount = () => {
        clientEvents.removeListener('EDeleteClicked', this.deleteClient);
        clientEvents.removeListener('EEditeClicked', this.showEditCard);
        clientEvents.removeListener('ESaveClicked', this.editClient);
        clientEvents.removeListener('ECancelClicked', this.hideCard);
        clientEvents.removeListener('EAddClicked', this.createNewClient);
    }

    createNewClient = (inputSurnameRefVal, inputNameRefVal, inputSecondNameRefVal, inputBalanceRefVal) => {

        let newClients = [...this.state.clients];
        
        let client = {};
        client.info = {};
        
        client.id = (this.state.lastId + 1);
        client.info.fam = inputSurnameRefVal;
        client.info.im = inputNameRefVal;
        client.info.otch = inputSecondNameRefVal;
        client.info.balance = inputBalanceRefVal;

        newClients.push(client);

        this.setState({
            clients: newClients,
            lastId: client.id,
        });
    }

    showNewCard = () => {
        let clientsArr = this.state.clients;
        let maxId = clientsArr[0].id;

        clientsArr.forEach( (el) => {
            if (el.id > maxId) maxId = el.id;
        });

        if(this.state.mode != 2) {
            this.setState({
                mode: 2,
                lastId: maxId,
                client: {},
            });
        }
    };

    editClient = (inputSurnameRefVal, inputNameRefVal, inputSecondNameRefVal, inputBalanceRefVal, id) => {

        let newClients = [...this.state.clients];
        newClients = newClients.forEach( (client, index) => {

            if(client.id === id) {

                let newClient = {...client};
                let newClientInfo = {...client.info};
                newClientInfo.fam = inputSurnameRefVal;
                newClientInfo.im = inputNameRefVal;
                newClientInfo.otch = inputSecondNameRefVal;
                newClientInfo.balance = inputBalanceRefVal;
                newClient.info = newClientInfo;
                newClients[index] = newClient;
            }

            this.setState({clients: newClients});

        })
    };
    
    hideCard = () => {
        this.setState({
            mode: null,
            client: {},
        });
    };

    showEditCard = (id) => {
        console.log("showEditCard сработала");
        let newClients = [...this.state.clients];
        let lookClient = newClients.filter( client => {
            return client.id === id;
        });

        lookClient = [...lookClient][0];

        if(this.state.mode != 1) {

            this.setState({
                clients: newClients,
                mode: 1,
                client: lookClient,
            });

        } else {

            this.setState({
                clients: newClients,
                client: lookClient,
            });
            
        }
    };

    deleteClient = (id) => {

        let newClients = [...this.state.clients];
        newClients.forEach((client, index) => {
            if (client.id === id) {
                newClients.splice(index, 1);
            }
        } );

        let maxId;

        if (newClients.length) {
            
            maxId = newClients[0].id;

        } else {

            maxId = 100;

        }

        newClients.forEach( (el) => {
            if (el.id > maxId) maxId = el.id;
        });

        this.setState({
            clients: newClients,
            lastId: maxId,
        });
    };

    setName_1 = () => {
        this.setState({name: 'Velcome'});
    };

    setName_2 = () => {
        this.setState({name: 'MTC'});
    };

    showActive = () => {
        let newClients; 
        let activeClients;

        if (!this.state.clientsCopy.length) {
            newClients = [...this.state.clients];
        } else {
            newClients = [...this.state.clientsCopy];
        }

        activeClients = newClients.filter(client => {
            return client.info.balance > 0;
        });

        this.setState({
            clients: activeClients,
            clientsCopy: newClients,
        });
    }

    showAll = () => {
        this.setState({
            clients: this.state.clientsCopy,
            clientsCopy: [],
        })
    }

    showNotActive = () => {
        let newClients; 
        let activeClients;
        
        if (!this.state.clientsCopy.length) {
            newClients = [...this.state.clients];
        } else {
            newClients = [...this.state.clientsCopy];
        }

        activeClients = newClients.filter(client => {
            return client.info.balance <= 0;
        });

        this.setState({
            clients: activeClients,
            clientsCopy: newClients,
        });
    }

    render() {
        console.log("MobileCompany render");

        let clientsCode = this.state.clients.map( client => {
            return <MobileClient key={client.id} id={client.id} info={client.info}/>
        });

        return (
            <div className="MobileCompany">
                <input type="button" value="Velcome" onClick={this.setName_1} />
                <input type="button" value="MTC" onClick={this.setName_2} />
                <div>
                    Компания: {this.state.name}
                </div>
                <input type="button" value="Все" onClick={this.showAll}/>
                <input type="button" value="Активные" onClick={this.showActive}/>
                <input type="button" value="Заблокированные" onClick={this.showNotActive}/>
                <table className="MobileCompany__table">
                    <tbody>
                        <tr>
                            <th>Фамилия</th>
                            <th>Имя</th>
                            <th>Отчество</th>
                            <th>Баланс</th>
                            <th>Статус</th>
                            <th>Редактировать</th>
                            <th>Удалить</th>
                        </tr>
                        {clientsCode}
                    </tbody>
                </table>
                <input
                    type="button"
                    value="Добавить клиента"
                    onClick={this.showNewCard}
                />
                {
                    (this.state.mode == 1) &&
                    <MobileClientCard 
                        mode={this.state.mode}
                        client={this.state.client}
                        key={this.state.client.id}
                    />
                }
                {
                    (this.state.mode == 2) &&
                    <MobileClientCard 
                        mode={this.state.mode}
                        lastId={this.state.lastId}
                        key={this.state.lastId + 1}
                    />
                }
            </div>
        );
    }
}

export default MobileCompany;