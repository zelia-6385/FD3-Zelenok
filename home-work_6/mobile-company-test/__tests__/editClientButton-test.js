"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';

test('работа editClientButton', () => {

    let clientArr = [
        {id:101, info: {fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}},
        {id:102, info: {fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:230}},
        {id:103, info: {fam:"Петров", im:"Петр", otch:"Петрович", balance:180}},
        {id:104, info: {fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:-280}},
    ];

    const component = renderer.create(
        <MobileCompany
        clients={clientArr}/>
    );

    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    const editClientButtons = component.root.findAll( el => el.props.value === 'Редактировать');

    // editClientButtons.forEach( elem => {
    //     elem.props.onClick();

    //     componentTree = component.toJSON();
    //     expect(componentTree).toMatchSnapshot();

    // });

    editClientButtons[0].props.onClick();

    componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
    
});