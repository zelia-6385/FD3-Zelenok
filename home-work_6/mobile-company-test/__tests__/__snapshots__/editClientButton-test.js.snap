// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа editClientButton 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table
    className="MobileCompany__table"
  >
    <tbody>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          230
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -280
        </td>
        <td
          className="MobileClient_red"
        >
          not active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`работа editClientButton 2`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table
    className="MobileCompany__table"
  >
    <tbody>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          230
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="MobileClient_green"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -280
        </td>
        <td
          className="MobileClient_red"
        >
          not active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
  <div
    className="MobileClientCard"
  >
    <div
      className="MobileClientCard__title"
    >
      Edit existing Client 101
    </div>
    <div
      className="MobileClientCard__row"
    >
      <label
        htmlFor="client-surname"
      >
        Surname: 
      </label>
      <input
        defaultValue="Иванов"
        id="client-surname"
        type="text"
      />
    </div>
    <div
      className="MobileClientCard__row"
    >
      <label
        htmlFor="client-name"
      >
        Name: 
      </label>
      <input
        defaultValue="Иван"
        id="client-name"
        type="text"
      />
    </div>
    <div
      className="MobileClientCard__row"
    >
      <label
        htmlFor="client-second-name"
      >
        Second Name: 
      </label>
      <input
        defaultValue="Иванович"
        id="client-second-name"
        type="text"
      />
    </div>
    <div
      className="MobileClientCard__row"
    >
      <label
        htmlFor="client-balance"
      >
        Balance: 
      </label>
      <input
        defaultValue={200}
        id="client-balance"
        type="text"
      />
    </div>
    <div
      className="MobileClientCard__controls"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Сохранить
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Отмена
      </button>
    </div>
  </div>
</div>
`;
